Design: Add CFP design folder and template (#23690)
cilium-team/cilium: design-cfps/README.md

> This repo contains CFP design proposals for features impacting repos across the

Reviewer: ```suggestion
This directory contains CFP design proposals for features impacting repos across the
```

Cilium Github organization.

# Purpose of CFPs

The purpose of a CFP is to allow community members to gain feedback
on their designs from the Committers before the community member commits to
executing on the design. By going through the design process, developers gain a
have a high level of confidence that their designs are viable and will be
accepted.

NOTE: This is process is not mandatory. Anyone can execute on their own design
without going through this process and submit code to the respective repos.
However, depending on the complexity of the design and how experienced the
developer is within the community, they could greatly benefit from going through
this design process first. The risk of not getting a design proposal approved
is that a developer may not arrive at a viable architecture that the community will
accept.

# How to create CFPs

To create a CFP, it is recommended to use the `CFP-YYYY-MM-DD-template.md`
file as an outline. The structure of this template is meant to provide a starting
point for people. Feel free to edit and modify your outline to best fit your

> needs when creating a proposal.

Reviewer: Maybe something on the naming format we expect? Are we going to go with "have an issue first, and name this after the issue number" like KEPs etc, or should we do `CFP-YYYY-MM-DD-subject.md`, where the date is the day the discussion is opened? I lean towards the former (it makes things more trackable), but not strongly.

Author: All of the CFPs I've seen so far use the latter format so I'll stick with that for now unless we decide something else.

> To create a CFP, it is recommended to use the `CFP-YYYY-MM-DD-template.md`
> file as an outline. The structure of this template is meant to provide a starting
> point for people. Feel free to edit and modify your outline to best fit your
> needs when creating a proposal.

Author: ```suggestion
To create a CFP, it is recommended to use the `CFP-YYYY-MM-DD-template.md`
file as an outline. The structure of this template is meant to provide a starting
point for people. Feel free to edit and modify your outline to best fit your
needs when creating a proposal. The title should be `CFP-YYYY-MM-DD-subject.md` 
where the date is the day the discussion opened.
```

This directory contains CFP design proposals for features impacting repos across the
Cilium Github organization.

# Purpose of CFPs

> The purpose of a CFP is to allow community members to gain feedback

Reviewer 2: I don't think you ever explain what CFP means?

```suggestion
The purpose of a Cilium Feature Proposal (CFP) is to allow community members to gain feedback
```

on their designs from the Committers before the community member commits to
executing on the design. By going through the design process, developers gain a

> have a high level of confidence that their designs are viable and will be

Reviewer 2: ```suggestion
executing on the design. By going through the design process, developers gain a
high level of confidence that their designs are viable and will be
```

accepted.

> NOTE: This is process is not mandatory. Anyone can execute on their own design

Reviewer 2: ```suggestion
NOTE: This process is not mandatory. Anyone can execute on their own design
```

without going through this process and submit code to the respective repos.
However, depending on the complexity of the design and how experienced the
developer is within the community, they could greatly benefit from going through
this design process first. The risk of not getting a design proposal approved
is that a developer may not arrive at a viable architecture that the community will
accept.

# How to create CFPs

To create a CFP, it is recommended to use the `CFP-YYYY-MM-DD-template.md`
file as an outline. The structure of this template is meant to provide a starting
point for people. Feel free to edit and modify your outline to best fit your

> needs when creating a proposal. The title should be `CFP-YYYY-MM-DD-subject.md` 

Reviewer 3: I wonder if this is the right time also to switch to just a numbered CFP? Something like, open a PR against the repo, then take the PR number as the `CFP-XXXX-subject.md`? Given that we tend to batch quite a bit of planning, it's not uncommon for the day to be the same for two proposals. It's then a bit odd to have a YYYY-MM-DD structure since the author has to pick another date.

Author: @youngnick also had this suggestion above, but I said to keep it the same, but I'm also find to switch

where the date is the day the discussion opened.

Many design docs also begin their life as a Google doc or other shareable
file for easy commenting and editing when still in the early stages of discussion.
Once your proposal is done, submit it as a PR to the design-cfps folder.

If you want to bring further attention to your design, you may want to

> raise the design during the weekly community call and on the [#development

Reviewer 2: Link to community meeting details?

channel in Slack](https://cilium.slack.com/archives/C2B917YHE).

# Getting a design approved

For a CFP to be considered viable, a CODEOWNER from each area impacted by
the design needs to provide an approval. Once all the approvals are collected,
the design can be merged. A merged design proposal means the proposal is

> viable to be executed on.

Reviewer 3: I'm fine with this, but we will also need to figure out the process to identify the relevant CODEOWNERS for any given CFP.

Author: Yes, is there a good way to do that or is there a better way you can think of doing it?

> viable to be executed on.

Reviewer 2: Could you elaborate a little on that part? In particular, what does it mean from Cilium's reviewers/committers that a CFP is merged: Do they commit to accept the design eventually (I guess not)? Or should we remind that even if there's a very high chance the implementation gets accepted, it's still not a 100% chance?
